File: 03 March/03-07/elementSize.js
function getElementSize(windowSize, elementVw, elementVh) {
  const { windowWidth, windowHeight } = /(?<windowWidth>\d+) x (?<windowHeight>\d+)/.exec(windowSize).groups;
  const { viewportWidthUnits } = /(?<viewportWidthUnits>\d+)vw/.exec(elementVw).groups;
  const { viewportHeightUnits } = /(?<viewportHeightUnits>\d+)vh/.exec(elementVh).groups;

  return `${windowWidth * viewportWidthUnits / 100} x ${windowHeight * viewportHeightUnits / 100}`;
}

// --- TEST SUITE ---

const testsText = String.raw`
1. getElementSize("1200 x 800", "50vw", "50vh") should return "600 x 400".
2. getElementSize("320 x 480", "25vw", "50vh") should return "80 x 240".
3. getElementSize("1000 x 500", "7vw", "3vh") should return "70 x 15".
4. getElementSize("1920 x 1080", "95vw", "100vh") should return "1824 x 1080".
5. getElementSize("1200 x 800", "0vw", "0vh") should return "0 x 0".
6. getElementSize("1440 x 900", "100vw", "114vh") should return "1440 x 1026".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);

File: 03 March/03-07/element_size.py
def get_element_size(window_size, element_vw, element_vh):
    window_dict = re.match(r"(?P<window_width>\d+) x (?P<window_height>\d+)", window_size).groupdict()
    viewport_width_dict = re.match(r"(?P<viewport_width_units>\d+)vw", element_vw).groupdict()
    viewport_height_dict = re.match(r"(?P<viewport_height_units>\d+)vh", element_vh).groupdict()

    return f"{int(window_dict['window_width']) * int(viewport_width_dict['viewport_width_units']) // 100} x {int(window_dict['window_height']) * int(viewport_height_dict['viewport_height_units']) // 100}"

# --- TEST SUITE ---

tests_text = r'''
1. get_element_size("1200 x 800", "50vw", "50vh") should return "600 x 400".
2. get_element_size("320 x 480", "25vw", "50vh") should return "80 x 240".
3. get_element_size("1000 x 500", "7vw", "3vh") should return "70 x 15".
4. get_element_size("1920 x 1080", "95vw", "100vh") should return "1824 x 1080".
5. get_element_size("1200 x 800", "0vw", "0vh") should return "0 x 0".
6. get_element_size("1440 x 900", "100vw", "114vh") should return "1440 x 1026".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)